ops/lost-in-translation.js

```diff
@@ -0,0 +1,163 @@
+ * This script finds all the english translation keys across
+ * all commits (including branches and pull requests) for each
+ * locale.json in app/locales. Afterwards it looks at each locale
+ * separately to see if it's missing any one the english keys.
+ * Usage: node ops/lost-in-translation.js
+ * Important: You need to run it from the root directory.
+ */
+
+/**
+ * Calls git cli with arguments and returns stdout.
+ */
+function git(...args) {
+  const { stdout } = spawnSync('git', args, { encoding: 'utf-8'});
+  return stdout;
+}
+
+/**
+ * Returns the commit hashes for a file path across all local branches.
+ */
+function findCommitHashesForFile(filePath) {
+  const output = git('log', '--pretty=format:"%h"', '--all', '--', filePath);
+  const hashes = output.split('\n').map(line => line.replace(/"/g, ''));
+  return hashes;
+}
+
+/**
+ * Returns the JSON representation of the contents of a file at certain commit hash.
+ */
+function retrieveJSONForFileAtCommitHash(filePath, commitHash) {
+  const contentsAtCommitHash = git('show', `${commitHash}:${filePath}`);
+  try {
+    return JSON.parse(contentsAtCommitHash);
+  } catch (error) {
+    // Sometimes the file is not in a proper JSON format. Simply return nothing in that case.
+    return {};
+  }
+}
+
+/**
+ * Normalizes translation key like we do in app/server.ts.
+ */
+function normalizeTranslationKey(translationKey) {
+  return translationKey.replace(/[\s\n\t]+/g, ' ').trim();
+}
+
+/**
+ * Find all the locales (ie. en-IN, no, se) in the provided directory.
+ */
+function retrieveAllLocales(directoryPath) {
+  const filenames = readdirSync(directoryPath);
+  const locales = filenames.map(filename => parse(filename).name);
+  return locales;
+}
+
+/**
+ * Add rows to a Google Spreadsheet that are not already added.
+ */
+async function addUniqueRowsToGoogleSheet(sheet, rowsToAdd) {
+  const alreadyAddedRows = await sheet.getRows({ limit: 1000 });
+  const uniqueRowsToAdd = rowsToAdd.filter(rowToAdd =>
+    !alreadyAddedRows.find(alreadyAddedRow => alreadyAddedRow.key === rowToAdd.key));
+  await sheet.addRows(uniqueRowsToAdd);
+  return uniqueRowsToAdd;
+}
+
+/**
+ * Step 1: Find all the (english) translation keys across all branches and PRs.
+ *
+ * If a Dutch developer has made a feature in a branch, we expect that him/her added a key
+ * to one or many locale files (usually the english locale (en.json) or to their own locale
+ * file (nl.json), or both).
+ *
+ * But they probably haven't added translations to all the other locales, because they don't
+ * speak the other languages. And this is the problem, because now we need find people to help
+ * translate the added keys to all the other locales.
+ *
+ * So what we do here is simply to find *all* the english translation keys across the entire
+ * project. That means looking at all the english translations keys in all the locale files
+ * since the start of the project across all branches and PRs. We throw them into a set that
```

Comment:
This _sounds_ unnecessary to me, as it would also include WIP's and stale branches. But it might be some cases where this makes sense that I have not thought about?

Reply:
Yeah, that's the part that sucks, but having it would make sure that *all* translations are added and (hopefully) translated by the time we want to merge it.

Reply:
Aha, I understand. So a possible solution would be to run this if this run as a cronjob or something, so that the sheet is updated as soon as there are new texts in the repo?

Reply:
Yeah, so in the case with Stano's PR, those translations are now added to the sheet (even though his PR is WIP). Not sure if we want that or not, but we could try it out?